sklearn/_loss/tests/test_loss.py

```diff
@@ -1356,3 +1369,92 @@ def test_tweedie_log_identity_consistency(p):
     assert_allclose(
         hessian_log, y_pred * gradient_identity + y_pred**2 * hessian_identity
     )
+
+
+@pytest.mark.parametrize(
+    "loss_class",
+    [HalfBinomialLoss, HalfMultinomialLoss],
+    ids=["HalfBinomialLoss", "HalfMultinomialLoss"],
+)
+@pytest.mark.parametrize(
+    "method_name", ["__call__", "gradient", "loss", "loss_gradient"]
+)
+@pytest.mark.parametrize("use_sample_weight", [False, True])
+@pytest.mark.parametrize(
+    "namespace, device_, dtype_name",
+    yield_namespace_device_dtype_combinations(),
+    ids=_get_namespace_device_dtype_ids,
+)
+def test_loss_array_api(
+    loss_class, method_name, use_sample_weight, namespace, device_, dtype_name
+):
+    def _assert_array_api_result(result_xp, result_np, raw_prediction_xp, xp, atol):
+        assert_allclose(_convert_to_numpy(result_xp, xp=xp), result_np, atol=atol)
+        assert result_xp.dtype == raw_prediction_xp.dtype
+        assert device(result_xp) == device(raw_prediction_xp)
+
+    xp = _array_api_for_tests(namespace, device_)
+    atol = _atol_for_type(dtype_name)
+    random_seed = 42
+    n_samples = 100
+    loss_instance = loss_class()
+    y_true, raw_prediction = random_y_true_raw_prediction(
+        loss=loss_instance,
+        n_samples=n_samples,
+        y_bound=(-100, 100),
+        raw_bound=(-5, 5),
```

Comment:
I think we might want to increase the width of this bound. The Cython code for `cgradient_half_binomial` or `log1pexp` has conditions for values below `-37` or above `33.3`, and I don't think we can cover those branches of the code with such constrained bounds.